File: Algorithm/Manacher's Algorithm/leetcode/B56/Claude/B56.py
# Python版での主な特徴と最適化ポイント：

# ## 🔧 型安全性の徹底

# 1. **完全な型注釈**: 全ての変数、パラメータ、戻り値に型を明示
# 2. **Pylance対応**: 静的型チェッカーでのエラー回避
# 3. **型ヒント活用**: `typing`モジュールの適切な使用

# ## ⚡ パフォーマンス最適化

# 1. **リスト内包表記回避**: メモリ効率を重視した明示的ループ
# 2. **文字列操作最適化**: `'#'.join(s)`による効率的な文字列結合
# 3. **インデックス計算**: 数値演算のみで高速処理

# ## 💾 メモリ効率化

# ```python
# # 効率的な配列初期化
# radius: List[int] = [0] * n  # リスト内包表記より高速

# # 不要なコピー回避
# processed: str = '#' + '#'.join(s) + '#'  # 一度だけ文字列生成
# ```

# ## 🎯 型エラー対策

# 1. **明示的型変換**: `map(int, ...)`の結果を適切に処理
# 2. **タプルの型指定**: クエリデータの構造化
# 3. **戻り値型の明示**: 関数の意図を明確化

# ## 📊 計算量分析

# - **時間計算量**: O(N + Q)
#   - Manacher's algorithm: O(N)
#   - クエリ処理: O(Q) × O(1) = O(Q)
# - **空間計算量**: O(N)
#   - 半径配列: O(N)
#   - 処理済み文字列: O(N)

# ## 🛡️ エラーハンドリング

# TypeScriptと同様の堅牢性を保ちながら、Python特有の最適化（リスト操作、文字列処理）を活用。Pylanceによる静的解析にも対応した型安全なコードです。

# この実装により、Python環境でも制約条件下で効率的に動作し、メモリ使用量を最小限に抑えながら高速な回文判定が可能です。

from typing import List, Tuple
import sys


def manacher(s: str) -> List[int]:
    """
    Manacher's algorithmを使用して回文の半径配列を計算
    
    Args:
        s (str): 入力文字列
        
    Returns:
        List[int]: 各位置での最長回文の半径を格納した配列
    """
    # 文字間に特殊文字を挿入して奇数長にする
    processed: str = '#' + '#'.join(s) + '#'
    n: int = len(processed)
    radius: List[int] = [0] * n
    center: int = 0
    right: int = 0
    
    for i in range(n):
        # 対称性を利用して初期値を設定
        if i < right:
            radius[i] = min(right - i, radius[2 * center - i])
        
        # 回文を拡張
        while (i + radius[i] + 1 < n and 
               i - radius[i] - 1 >= 0 and 
               processed[i + radius[i] + 1] == processed[i - radius[i] - 1]):
            radius[i] += 1
        
        # centerとrightを更新
        if i + radius[i] > right:
            center = i
            right = i + radius[i]
    
    return radius


def is_palindrome(radius: List[int], l: int, r: int) -> bool:
    """
    指定された範囲が回文かどうかを判定
    
    Args:
        radius (List[int]): Manacher's algorithmで計算された半径配列
        l (int): 開始位置（1-indexed）
        r (int): 終了位置（1-indexed）
        
    Returns:
        bool: 回文の場合True、そうでなければFalse
    """
    # 1-indexedを0-indexedに変換
    start_idx: int = l - 1
    end_idx: int = r - 1
    
    # 処理済み文字列での位置計算
    center: int = start_idx + end_idx + 1  # 中心位置
    length: int = end_idx - start_idx + 1  # 部分文字列の長さ
    
    # 必要な半径と実際の半径を比較
    return radius[center] >= length


def parse_input(input_data: str) -> Tuple[int, int, str, List[Tuple[int, int]]]:
    """
    入力データをパース
    
    Args:
        input_data (str): 標準入力の内容
        
    Returns:
        Tuple[int, int, str, List[Tuple[int, int]]]: (N, Q, S, クエリリスト)
    """
    lines: List[str] = input_data.strip().split('\n')
    n, q = map(int, lines[0].split())
    s: str = lines[1]
    
    queries: List[Tuple[int, int]] = []
    for i in range(q):
        l, r = map(int, lines[2 + i].split())
        queries.append((l, r))
    
    return n, q, s, queries


def solve(input_data: str) -> str:
    """
    メイン処理関数
    
    Args:
        input_data (str): 標準入力の内容
        
    Returns:
        str: 結果の出力文字列
    """
    # n, q, s, queries = parse_input(input_data)
    _, _, s, queries = parse_input(input_data)
    
    # Manacher's algorithmで事前計算
    radius: List[int] = manacher(s)
    
    results: List[str] = []
    
    # 各クエリを処理
    for l, r in queries:
        if is_palindrome(radius, l, r):
            results.append('Yes')
        else:
            results.append('No')
    
    return '\n'.join(results)


def main() -> None:
    """
    エントリーポイント
    標準入力から読み込み、結果を標準出力に出力
    """
    # メモリ効率を考慮して一度に読み込み
    input_data: str = sys.stdin.read()
    result: str = solve(input_data)
    print(result)


if __name__ == '__main__':
    main()

File: Algorithm/Manacher's Algorithm/leetcode/B56/Claude/B56.ts
// TypeScript版での主な改善点と特徴：

// ## 型安全性の向上

// 1. **厳密な型定義**: 全ての変数、パラメータ、戻り値に型を明示
// 2. **インターフェース定義**: `Query`と`InputData`で構造化されたデータ型
// 3. **型推論の活用**: TypeScriptの型推論機能を適切に利用

// ## メモリ効率化の考慮

// 1. **適切な配列初期化**: `new Array(n).fill(0)`でメモリを効率的に確保
// 2. **不要なオブジェクト生成を避ける**: プリミティブ型を優先使用
// 3. **文字列操作の最適化**: 必要最小限の文字列処理

// ## パフォーマンス最適化

// 1. **事前計算**: Manacher's algorithmでO(N)時間での前処理
// 2. **クエリ処理**: O(1)時間での高速回文判定
// 3. **メモリアクセス最適化**: 配列アクセスパターンを考慮

// ## 型定義の詳細

// - `Query`: クエリの開始・終了位置を格納
// - `InputData`: 入力全体の構造化表現
// - 関数の型注釈により、入出力の型を明確化

// ## 計算量
// - **時間計算量**: O(N + Q)
// - **空間計算量**: O(N)

// この実装により、TypeScript 5.1環境でも制約条件下で効率的に動作し、型安全性を保ちながら高速な回文判定が可能です。

import * as fs from 'fs';

/**
 * Manacher's algorithmを使用して回文の半径配列を計算
 * @param s - 入力文字列
 * @returns 各位置での最長回文の半径を格納した配列
 */
function manacher(s: string): number[] {
    // 文字間に特殊文字を挿入して奇数長にする
    const processed: string = '#' + s.split('').join('#') + '#';
    const n: number = processed.length;
    const radius: number[] = new Array(n).fill(0);
    let center: number = 0;
    let right: number = 0;

    for (let i: number = 0; i < n; i++) {
        // 対称性を利用して初期値を設定
        if (i < right) {
            radius[i] = Math.min(right - i, radius[2 * center - i]);
        }

        // 回文を拡張
        while (
            i + radius[i] + 1 < n &&
            i - radius[i] - 1 >= 0 &&
            processed[i + radius[i] + 1] === processed[i - radius[i] - 1]
        ) {
            radius[i]++;
        }

        // centerとrightを更新
        if (i + radius[i] > right) {
            center = i;
            right = i + radius[i];
        }
    }

    return radius;
}

/**
 * 指定された範囲が回文かどうかを判定
 * @param radius - Manacher's algorithmで計算された半径配列
 * @param l - 開始位置（1-indexed）
 * @param r - 終了位置（1-indexed）
 * @returns 回文の場合true、そうでなければfalse
 */
function isPalindrome(radius: number[], l: number, r: number): boolean {
    // 1-indexedを0-indexedに変換
    const startIdx: number = l - 1;
    const endIdx: number = r - 1;

    // 処理済み文字列での位置計算
    const center: number = startIdx + endIdx + 1; // 中心位置
    const len: number = endIdx - startIdx + 1; // 部分文字列の長さ

    // 必要な半径と実際の半径を比較
    return radius[center] >= len;
}

/**
 * クエリ情報の型定義
 */
interface Query {
    l: number;
    r: number;
}

/**
 * 入力データの型定義
 */
interface InputData {
    n: number;
    q: number;
    s: string;
    queries: Query[];
}

/**
 * 入力データをパース
 * @param input - 標準入力の内容
 * @returns パースされた入力データ
 */
function parseInput(input: string): InputData {
    const lines: string[] = input.trim().split('\n');
    const [n, q]: number[] = lines[0].split(' ').map(Number);
    const s: string = lines[1];

    const queries: Query[] = [];
    for (let i: number = 0; i < q; i++) {
        const [l, r]: number[] = lines[2 + i].split(' ').map(Number);
        queries.push({ l, r });
    }

    return { n, q, s, queries };
}

/**
 * メイン処理関数
 * @param input - 標準入力の内容
 * @returns 結果の出力文字列
 */
function solve(input: string): string {
    const data: InputData = parseInput(input);

    // Manacher's algorithmで事前計算
    const radius: number[] = manacher(data.s);

    const results: string[] = [];

    // 各クエリを処理
    for (const query of data.queries) {
        if (isPalindrome(radius, query.l, query.r)) {
            results.push('Yes');
        } else {
            results.push('No');
        }
    }

    return results.join('\n');
}

// 標準入力から読み込み
const input: string = fs.readFileSync('/dev/stdin', 'utf8');
console.log(solve(input));
